### PR TITLE
Replacement for Base.Test (ready for review)

### PR DESCRIPTION
**Pull request by IainNZ**
The goal of this PR is to make `Base.Test` better than just `assert` with some fancier output. It does so primarily by introducing `@testset`, which lets users bundle tests together and delay throwing an error until the end.

Design-wise, it primarily needs to be non-breaking for 99.9% uses of the current `Base.Test`, and be simple, minimal, and relatively opinion-free. It is clear there is a large space of possibilities for testing frameworks, but I don't want to go there (personally). My goal is for a incremental improvement over the existing `Base.Test` - if someone wants to pick up the ball and add more features later, they can. A thing that falls outside the scope includes fancier outputs (to, e.g. files).

This PR is basically https://github.com/IainNZ/BaseTestNext.jl. It is a drop-in replacement for `Base.Test`, and all tests except for those in `test/test.jl` pass without modification.

Features for later PRs (probably not by me):
- Use the special formatting on Travis for expand/collapse (via @simonbyrne)
- Enable changing the default test set used, which then enables custom output (e.g. JSON, XML)
- `@test_approx` needs to be either included in the testset world, or even better deprecated in favor of `isapprox` based testing.

Some demo output from BaseTestNext.jl

Imgur

### CODE REVIEW

**Comment by jakebolewski** [comment]
:100:

**Comment by IainNZ** [comment]
@tkelman any ideas why `mmap` is dying on Windows? As far as I can, its the only thing this PR breaks, but the tests for `mmap` are somewhat impenetrable to me. The line in the backtrace is particularly as there isn't even a test there!

File: test/core.jl [hunk 1]
```diff
@@ -3100,15 +3100,25 @@ Base.convert(::Type{Foo11874},x::Int) = float(x)
 
 # issue #9233
 let
-    err = @test_throws TypeError NTuple{Int, 1}
```
Comment (tkelman) [inline test/core.jl]: you can't make `@test_throws` return the error?

Comment (IainNZ) [inline test/core.jl]: It'd be inconsistent with the `@test`. Now everything returns a test object, with nice behaviour at the REPL as a result. I think using `@test_throws` as a one-liner try-catch was not a good idea to begin with, although thats completed unrelated to why I picked the return type.

Comment (tkelman) [inline test/core.jl]: okay fair enough, sounds like there's a good reason for the change. slightly breaking if any packages were doing this, but probably worth it

**Comment by aviks** [comment]
+100. This is the one thing I've wanted as far as incremental improvements to Base.Test go.

**Comment by ScottPJones** [comment]
:+1: :100: I'd lobbied for just this sort of thing a while ago, great to see something like this actually being built.

**Comment by tbreloff** [comment]
This is exciting... do you see it making FactCheck.jl obsolete at some
point?

**Comment by IainNZ** [comment]
@tbreloff it makes FactCheck obsolete for my personal needs, and I won't be working on it any more, but as a package it is always going to have much more flexibility in how it does things because it can be quite opinionated/crazy if it wants.

**Comment by quinnj** [comment]
@IainNZ, my guess is the `mmap` failure probably has to do with trying to mmap an open file, which results in the invalid argument. We may need to explicitly call `finalize` more on the mmapp-ed arrays to ensure the files are getting garbage collected.

**Comment by IainNZ** [comment]
@quinnj Yes that makes sense. This shouldn't be keeping any references around any longer than the old `@test` did, but it would probably change the GC pattern enough to reveal something. If you have any ideas on how to fix (or make the tests more robust), I'm all ears!

**Comment by quinnj** [comment]
Just pushed some changes that make mmap tests pass on my windows machine.

**Comment by quinnj** [comment]
(to this PR)

**Comment by IainNZ** [comment]
Hah, well I'll be damned, very nice work.
I'm going to rebase the commits to put that first, then the rest of it on top of that so we have a passing history.

**Comment by IainNZ** [comment]
Ok, lets see how CI likes that

**Comment by ssfrr** [comment]
To quote from the #13090:

with `@test_throws` returning an exception:
```julia
err = @test_throws TypeError NTuple{Int, 1}
@test err.func == :NTuple
@test err.expected == Int
@test err.got == Int
```

becomes (if `@test_throws` doesn't return the exception):

```julia
try
    NTuple{Int, 1}
    @test false
catch err
    @test isa(err, TypeError)
    @test err.func == :NTuple
    @test err.expected == Int
    @test err.got == Int
end
```

I think the first is more clear, and the extra `try...catch` stuff sticks out even more if you're only checking one property of the exception (say `err.msg`). I really like @tkelman's [idea](https://github.com/JuliaLang/julia/pull/13090#issuecomment-139794445) of attaching the exception to the `Result`, seems like the best of both worlds.

**Comment by IainNZ** [comment]
I don't think the first is more clear, I think its using `@test_throws` for more than its intended. But returned the exception as part of the result type is fine, and I'll tweak it so it does (it currently only does so if the wrong test type was thrown, so its a 5 character change).

**Comment by IainNZ** [comment]
OK, now returns the exception value always (not just on failure)

**Comment by IainNZ** [comment]
Now #13090 is in, rebased.

**Comment by milktrader** [comment]
Cool work @IainNZ

**Comment by tkelman** [comment]
I guess anywhere in the testing docs that mentions handlers should probably be updated here since that mechanism is being removed.

**Comment by IainNZ** [comment]
Yeah I pulled that out of the helpdb thing to get it to compile, but I need to substantially rewrite that chapter before merging this.

**Comment by IainNZ** [comment]
So I'm trying to do something sensible with `test_approx_eq`, but they don't really work like any of the rest of the tests anyway. I noticed that
```
macro test_approx_eq_eps(a, b, c)
    ex = :(isapprox($a, $b, atol=$c))
    :(do_test(()->($(Expr(:quote,ex)), $(esc(ex))), $(Expr(:quote,ex))))
end
```
actually passes all the tests in base, but that just might be because its easier to satisfy than the current criteria (it is essentially `maxabs` now, I think, and `isapprox` does `vecnorm`).

Is punting on "integrating" these into the new framework a viable option? I'm running out of time and steam for this PR.

**Comment by StefanKarpinski** [comment]
Yes, punting on these is very reasonable. I'd love to get rid of them altogether. They should be normal tests with a different predicate, not some weird special testing macro.

**Comment by stevengj** [comment]
@IainNZ, if you want to use the more-stringent `maxabs` check, you can pass `norm=maxabs` (or `norm=x->vecnorm(x,Inf)`) to `isapprox`.   Though maybe `isapprox` will need to be modified to take a `norm` keyword argument in the scalar case first.

**Comment by IainNZ** [comment]
@stevengj that problem with the scalar case is indeed what I hit

File: base/interactiveutil.jl [hunk 2]
```diff
@@ -10,39 +17,37 @@ function edit(file::AbstractString, line::Integer)
     else
         default_editor = "emacs"
     end
-    editor = get(ENV,"JULIA_EDITOR", get(ENV,"VISUAL", get(ENV,"EDITOR", default_editor)))
```
Comment (tkelman) [inline base/interactiveutil.jl]: this looks unrelated

Comment (IainNZ) [inline base/interactiveutil.jl]: somethings wrong, I've got another commit in there!

Comment (IainNZ) [inline base/interactiveutil.jl]: See that first one?

Comment (IainNZ) [inline base/interactiveutil.jl]: Rebased on master one more time, force pushed again, seems ok now.

File: doc/stdlib/test.rst [hunk 3]
```diff
@@ -1,83 +1,188 @@
...
+The ``Base.Test`` module provides simple *unit testing* functionality.
+Unit testing is a way to see if your code is correct by checking that
+the results are what you expect. It can be helpful to ensure your code
+still works after you make changes, and can be used when developing as
+a way of specifying the behaviors your code should have when complete.
 
-Julia is under rapid development and has an extensive test suite to
-verify functionality across multiple platforms. If you build Julia
-from source, you can run this test suite with ``make test``. In a
-binary install, you can run the test suite using ``Base.runtests()``.
```
Comment (tkelman) [inline doc/stdlib/test.rst]: :(

Comment (IainNZ) [inline doc/stdlib/test.rst]: Ah yes, I nuked this, but I think it should go somewhere else that isn't the docs (it was an awkward fit here). I'm thinking README.md

Comment (tkelman) [inline doc/stdlib/test.rst]: It's an unexported but documented function.

Comment (IainNZ) [inline doc/stdlib/test.rst]: Devdocs? "Essentials"? Anywhere but here would make just as much if not more sense to me.

Comment (tkelman) [inline doc/stdlib/test.rst]: It's about testing, not strictly info for contributors, and too specific for essentials imo. It isn't about Base.Test but i think the docs belong here in a separate section header regardless.

Comment (IainNZ) [inline doc/stdlib/test.rst]: OK, I'll put it back

**Comment by IainNZ** [comment]
OK, needs a final look before merging

**Comment by ssfrr** [comment]
Is it expected that passed testsets are displayed if their sibling fails? e.g.

```julia
@testset "outer" begin # should be displayed
    @testset "inner 1" begin # should not be displayed?
        @test true
    end
    @testset "inner 2" begin # should be displayed
        @test false
    end
end
```
gives
```
inner 2: Test Failed
  Expression: false
   Evaluated: false
 in record at /Users/srussell/.julia/v0.4/BaseTestNext/src/defaulttestset.jl:24
 in do_test at /Users/srussell/.julia/v0.4/BaseTestNext/src/BaseTestNext.jl:188
Test Summary: | Pass  Fail  Total
outer         |    1     1      2
  inner 1     |    1            1
  inner 2     |          1      1
ERROR: Some tests did not pass: 1 passed, 1 failed, 0 errored.
 in finish at /Users/srussell/.julia/v0.4/BaseTestNext/src/defaulttestset.jl:89
```

I don't really have a preference one way or the other (it could provide some extra context), but the docs say:


> In the event that a nested test set has no failures, as happened here, it will be hidden in the summary. If we do have a test failure, only the details for the failed test sets will be shown.

which makes it sound like it might be a bug.

**Comment by ssfrr** [comment]
BTW, I don't think that has to be a blocking issue even if it's not intended behavior. I know you're trying to tie a bow on this thing so if it merges as-is and we file an issue, I can look into it. Or just change the docs. :)

**Comment by KristofferC** [comment]
Great!

**Comment by stevengj** [comment]
Needs a 0.5 NEWS entry...

**Comment by IainNZ** [comment]
Is backporting to 0.4 viable?

**Comment by IainNZ** [comment]
@ssfrr it is intentional, I guess I just view it differently. So if a testset has a failure, then that failure is also associated with its parent. If I hid the "sibling" test sets it looks a bit confusing (especially when their a multiple failures), because the "children" don't sum up to the parent's totals

**Comment by stevengj** [comment]
Backporting a new feature like this doesn't make sense to me.

**Comment by IainNZ** [comment]
Not even to 0.4.1?

**Comment by bicycle1885** [comment]
Is that possible to provide this as a separated package (or Compat.jl)? I really want to use it!

**Comment by StefanKarpinski** [comment]
It's a big feature, but not really a normal language feature. The question is what breaks if we change this?

**Comment by IainNZ** [comment]
Nothing breaks, unless you used `with_handler`.
According to 
https://github.com/search?l=julia&q=with_handler&type=Code&utf8=%E2%9C%93
Some people did use it

**Comment by IainNZ** [comment]
(it was mostly used for lack of an easy way to not stop on the first failing test, and a lot of the results in that search are forks of Julia)

**Comment by StefanKarpinski** [comment]
We can also still technically sneak features into 0.4 since it's not final yet. @IainNZ, how painful will maintaining support for two different Base testing APIs be?

**Comment by IainNZ** [comment]
I'll investigate today, I'm guessing not terrible.

**Comment by tkelman** [comment]
I'm very much against backporting anything into release-0.4 right now that isnt a critical bugfix or absolutely harmless (like doc additions). We've done 2 rc's already, now's not the time. Maybe for 0.4.1.

**Comment by StefanKarpinski** [comment]
Well, 0.4.1 isn't the time since technically this is a breaking change and that would violate our principle that your code should not break between 0.4.0 and 0.4.1; otoh, there's no such promise going from 0.4.0-rc2 to 0.4.0.

**Comment by tkelman** [comment]
We're trying to get a release out that's way behind schedule. This needs people testing and using it, it is several hundred lines of code after all. We're technically being stricter than semver since we're pre 1.0. The least risky way of supporting this on 0.4 is as a package.

**Comment by mbauman** [comment]
Can the changes to `with_handler` be made nonbreaking?  Then it could be introduced purely as a feature for 0.4.1.  Still not ideal to introduce features in patch releases, but I think it'd be ok (especially since it'll have had a month or two of testing on master).

**Comment by ssfrr** [comment]
the other breaking change is the return value of `@test_throws`, which now returns a `Result` instead of the raw exception.

**Comment by IainNZ** [comment]
@ssfrr that isn't a breaking change anymore, as it turned out that feature was added in 0.4 and has since been removed by this PR.

@mbauman probably.

My initial plan was to have this is a package for all of 0.4 anyway, so getting it into 0.4.0 isn't personally critical to me.
